presto-main/src/main/java/com/facebook/presto/metadata/AllNodes.java

```diff
@@ -43,13 +45,20 @@ public AllNodes(
         this.activeCoordinators = ImmutableSet.copyOf(requireNonNull(activeCoordinators, "activeCoordinators is null"));
         this.activeResourceManagers = ImmutableSet.copyOf(requireNonNull(activeResourceManagers, "activeResourceManagers is null"));
         this.activeCatalogServers = ImmutableSet.copyOf(requireNonNull(activeCatalogServers, "activeCatalogServers is null"));
+
+        this.activeWorkerCount = Sets.difference(Sets.difference(activeNodes, activeResourceManagers), activeCatalogServers).size();
```

Comment:
IIRC Catalog Servers were never completely flushed out and should be empty in prod.
You could simplify this with something like (and also get another perf win) by getting rid of the second `Sets.difference`  call since `Sets.difference` iterates across the entire length of set1 and checks if set2 contains the elem from set1.

```
Set<InternalNode> nonRMNodes = Sets.difference(activeNodes, activeResourceManagers);
if (activeCatalogServers.isEmpty()) {
   return nonRMNodes;   
}
else {
return Sets.difference(nonRMNodes, activeCatalogServers);
}
```

Reply:
Do not think we need that optimization